### PR TITLE
Feat/gan

### PR DESCRIPTION
**Pull request by AleAyotte**
1)RED-WGAN added
2)Pretrained VGG added

Work in progress not ready yet to merge with the master

### CODE REVIEW

File: network/models/wgan.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        Create a Encoder module from the paper (Ref 1)
+
+        :param fmap_in: Number of input feature maps
+        :param fmap_out: Number of output features maps (Default: 2*fmap_in)
+        :param kernel: Kernel size of the convolution (Default: 3)
+        :param stride: Stride of the convolution (Default: 1)
+        :param dilation: Space size between each kernel. (Default=1)
+        :param padding: Padding size of the convolution (Default: 3)
+        """
+        super(RedEncoder, self).__init__()
```
Comment (boiteaclou) [inline network/models/wgan.py]: Old call to super, you can replace by `super().__init__()`

File: network/models/wgan.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+
+class RedEncoder(nn.Module):
+    def __init__(self, fmap_in, fmap_out=None, kernel=3, stride=1, dilation=1, padding=1):
+        """
+        Create a Encoder module from the paper (Ref 1)
+
+        :param fmap_in: Number of input feature maps
+        :param fmap_out: Number of output features maps (Default: 2*fmap_in)
+        :param kernel: Kernel size of the convolution (Default: 3)
```
Comment (boiteaclou) [inline network/models/wgan.py]: Replace parameter name by `kernel_size` or `ker_size`

File: network/models/wgan.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        :param stride: Stride of the convolution (Default: 1)
+        :param dilation: Space size between each kernel. (Default=1)
+        :param padding: Padding size of the convolution (Default: 3)
+        """
+        super(RedEncoder, self).__init__()
+
+        if fmap_out is None:
+            fmap_out = 2 * fmap_in
+
+        self.conv = nn.Sequential(
```
Comment (boiteaclou) [inline network/models/wgan.py]: Replace attribute name by `self.layers` or something ...

File: network/models/wgan.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+
+class RedDecoder(nn.Module):
+    def __init__(self, fmap_in, fmap_out=None, kernel=3, stride=1, dilation=1, padding=1, bn=True):
+        """
+        Create a 3D Decoder module from the paper (Ref 1)
+
+        :param fmap_in: Number of input feature maps
+        :param fmap_out: Number of output features maps (Default: fmap_in / 2)
+        :param kernel: Kernel size of the convolution. (Default: 3)
```
Comment (boiteaclou) [inline network/models/wgan.py]: Same as for encoder

File: network/models/wgan.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        :param stride: Stride of the convolution (Default: 1)
+        :param dilation: Space size between each kernel. (Default=1)
+        :param padding: Padding size of the convolution (Default: 3)
+        :param bn: If true, a batch norm layer will before the leakyReLU.
+                   Else, no batchnorm and ReLU activation
+        """
+        super(RedDecoder, self).__init__()
```
Comment (boiteaclou) [inline network/models/wgan.py]: Same as for encoder

File: network/models/wgan.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+        if fmap_out is None:
+            fmap_out = int(fmap_in / 2)
+
+        self.deconv = nn.ConvTranspose3d(fmap_in, fmap_out, kernel_size=kernel, stride=stride,
```
Comment (boiteaclou) [inline network/models/wgan.py]: Something weird is happening here with the spaces/indentation. To fix

Comment (boiteaclou) [reply network/models/wgan.py]: Also, change the attribute name to `self.layers` or something

File: network/models/wgan.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                                         dilation=dilation, padding=padding)
+        if bn:
+            self.act = nn.Sequential(
```
Comment (boiteaclou) [inline network/models/wgan.py]: You can expand the name to `self.activation`

File: network/models/wgan.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                encoder_config[i, 0] = 2**(5 + i)  # 32, 64, 128, 256
+
+        self.num_block = len(encoder_config)
+        self.encoder = nn.ModuleList()
+        self.decoder = nn.ModuleList()
+
+        # We create the encoder and the decoder
+        for i in range(len(encoder_config)):
+            self.encoder.append(
+                RedEncoder(fmap_in=encoder_config[i-1, 0] if i > 0 else 1,
```
Comment (boiteaclou) [inline network/models/wgan.py]: Something weird is also happening here. Fix

**Comment by AleAyotte** [comment]
If you want to test it:

image_shape = np.array([10, 10, 10])
my_gen = wgan.RedGenerator()
my_disc = wgan.RedDiscriminator(image_shape)
out_gen = my_gen(features)
out_disc = my_disc(out_gen)

File: network/models/vgg.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+"""
+
+import torchvision.models as models
+
+
+def load_vgg_model(num_layer=11, bn=True, pretrained=False):
+    """
+
+    :param num_layer: Number of layer of the VGG model.
+                      (must be 11, 13, 16 or 19)
```
Comment (boiteaclou) [inline network/models/vgg.py]: There is still some bad docstrings here

File: network/models/wgan.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+        List of all enconder block
+    decoder : nn.ModuleList
+        List of all decoder block
+
+    Methods
+    -------
+    forward(x)
+        apply the forward pass of the model on the data x
+    """
+    def __init__(self, config=None):
```
Comment (boiteaclou) [inline network/models/wgan.py]: I've been thinking about it. I think i would prefer for the config arguments to be parameters of the init. You can then construct a numpy array from them of you want. It makes the interfaces of the classes way clearer.

**Pull request merged**